### PR TITLE
Group as context manager

### PR DESCRIPTION
**Pull request by clbarnes**
For compatibility with other group-like things (e.g. `h5py.File`),
`Groups` now have enter/exit methods for use as a context manager.
If the underlying store has a `close` method, that will be called on
exit; otherwise this does nothing.

See issue #509.

It would have been nice to extend the context manager stuff to a Store superclass (an ABC which also inherits from MutableMapping), but as described in the issue, that makes guarantees that existing code may already break (e.g. using a raw `dict` as a store).

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [ ] Docs build locally (e.g., run ``tox -e docs``)
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

### CODE REVIEW

**Comment by clbarnes** [comment]
`tox -e docs` fails due to numcodecs not being installed. Is this a known issue?

File: zarr/tests/test_hierarchy.py [hunk 1]
```diff
@@ -912,6 +919,17 @@ def create_store():
...
+    def test_context_manager(self):
+
+        with self.create_group() as g:
+            store = g.store
+            d = g.create_dataset('foo/bar', shape=100, chunks=10)
+            d[:] = np.arange(100)
+
+        error = ValueError if sys.version_info >= (3, 6) else RuntimeError
+        with pytest.raises(error):
+            store.zf.extractall()
```
Comment (alimanfoo) [inline zarr/tests/test_hierarchy.py]: Just checking, is this testing that the underlying ZipStore has been closed and hence the zip file is closed? If so, might be useful to add a comment that this is what this is doing.

**Comment by alimanfoo** [comment]
> `tox -e docs` fails due to numcodecs not being installed. Is this a known issue?

Hm, no, not sure what's happening there.

File: zarr/tests/test_hierarchy.py (hunk 1 above)
Comment (clbarnes) [reply zarr/tests/test_hierarchy.py]: Exactly. I'll do that.

**Pull request merged**